libraries/chain/include/eosio/chain/snapshot_detail.hpp

```diff
@@ -0,0 +1,226 @@
+         , additional_signatures(bs.additional_signatures)
+      {}
+   };
+
+   /**
+    *      Snapshot V7 Data structures
+    *      ---------------------------
+    */
+   struct snapshot_block_state_legacy_v7 : public snapshot_block_header_state_legacy_v3 {
+      // additional member that can be present in `Transition Legacy Block` and
+      // is needed to convert to `Transition IF Block` (see https://github.com/AntelopeIO/leap/issues/2080)
+      using valid_t = uint32_t; // snapshot todo
+      std::optional<valid_t> valid;
+
+      snapshot_block_state_legacy_v7() = default;
+
+      snapshot_block_state_legacy_v7(const block_state_legacy& bs)
+         : snapshot_block_header_state_legacy_v3(bs)
+         , valid(0)  // snapshot todo
+      {}
+   };
+
+   struct snapshot_block_state_v7 {
+      // from block_header_state
+      block_id_type                                       block_id;
+      block_header                                        header;
+      protocol_feature_activation_set_ptr                 activated_protocol_features;
+      finality_core                                       core;
+      incremental_merkle_tree                             proposal_mtree;
+      incremental_merkle_tree                             finality_mtree;
+      finalizer_policy_ptr                                active_finalizer_policy;
+      proposer_policy_ptr                                 active_proposer_policy;
+      flat_map<block_timestamp_type, proposer_policy_ptr> proposer_policies;
+      flat_map<uint32_t, finalizer_policy_ptr>            finalizer_policies;
+
+      // from block_state
+      using valid_t = uint32_t; // snapshot todo
+      std::optional<valid_t> valid;
+
+      snapshot_block_state_v7() = default;
+
+      snapshot_block_state_v7(const block_state& bs)
+         : block_id(bs.block_id)
+         , header(bs.header)
+         , activated_protocol_features(bs.activated_protocol_features)
+         , core(bs.core)
+         , proposal_mtree(bs.proposal_mtree)
+         , finality_mtree(bs.finality_mtree)
+         , active_finalizer_policy(bs.active_finalizer_policy)
+         , active_proposer_policy(bs.active_proposer_policy)
+         , proposer_policies(bs.proposer_policies)
+         , finalizer_policies(bs.finalizer_policies)
+         , valid(0)  // snapshot todo
+           // other members of block_state strong_digest, weak_digest, valid_qc? // snapshot todo
+      {}
+   };
+
+   struct snapshot_block_state_variant_v7 {
+      static constexpr uint32_t minimum_version = 7;
+      static constexpr uint32_t maximum_version = 7;
+
+      std::variant<snapshot_block_state_legacy_v7, snapshot_block_state_v7> v;
+
+      snapshot_block_state_variant_v7() = default;
+
+      snapshot_block_state_variant_v7(const block_state& bs)
+         : v(snapshot_block_state_v7(bs))
+      {}
+
+      snapshot_block_state_variant_v7(const block_state_legacy& bs)
```

Comment:
I personally prefer all one-arg constructors to be `explicit`.

Reply:
Changing it. Personally I don't see a huge benefit in this case where unexpected conversions are unlikely.